src/internal/connector/onedrive/collection.go

```diff
@@ -347,6 +368,41 @@ func downloadContent(
 	return content, nil
 }
 
+// readFromCache fetches latest download URL from the cache and attempts to
+// download the file using the new URL.
+func (oc *Collection) readFromCache(
+	ctx context.Context,
+	iaag itemAndAPIGetter,
+	itemID string,
+) (io.ReadCloser, error) {
+	if oc.cache == nil {
+		return nil, clues.New("nil cache")
+	}
+
+	props, err := oc.cache.getItemProperties(ctx, itemID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Handle newly deleted items
+	if props.isDeleted {
+		logger.Ctx(ctx).Info("item deleted in cache")
+
+		return nil, graph.ErrDeletedInFlight
+	}
+
+	rc, err := downloadFile(ctx, iaag, props.downloadURL)
```

Comment:
A download isn't a read from a cache, which makes this behavior opaque and unexpected since the func name sounds like you're only reading from a cache.

All in all, I don't think the Collection makes for a sensible owner here.  Presumably, this is a collection method because the receiver already contains the dependencies necessary to perform the logic.  But outside of that, there's little precedent for the Collection being the thing that owns item retrieval wholesale.  And, imo, it's a simpler world if we don't extend that ownership.

I'd recommend turning this into a standalone func like `readItemContents()` that accepts, in addition to its other parameters, the oc.cache.  That'll make testing more simple by minimizing coupling between the collection and this behavior.

Reply:
`readItemContents` sounds good. 

> Re: Recommendation on turning it into a standalone function

Hmm, I am torn on this. I see url cache as a property of the collection ( or drive, to be precise). My idea behind `readFromCache` was that it acts as the url cache client and reads from cache, if one is attached to a collection.

Passing the cache around as an argument just seems wrong for some reason. Question to ask is whether we'll ever access the url cache without having a collection(keeping aside tests). 

Decoupling cache from collection can make testing slightly simpler, since it narrows the scope and won't need us to define collections every time we test url cache. 

Let me think a bit more about it. I think either way is fine at the risk of over discussion. If we are going to use static, standalone pattern here, we might have to extend that towards other methods defined in `Collection`.

Reply:
> Question to ask is whether we'll ever access the url cache without having a collection(keeping aside tests).

mmm, I think that's the wrong question to ask.  The logical claim that question supposes would be: "if A contains B, and a function uses B, then the function should use A."  But the function does not necessarily use A.  It may only use B, and should be able to work independent of A.  By that same reasoning, any top-level client or controller must be passed down into every leaf function (or leaf functions should be methods on the controller), purely because the controller contains the things those leaf functions use. 

This isn't exactly [dependency inversion](https://stackify.com/dependency-inversion-principle/), but SOLID principles feels topical here at large.

Reply:
Thanks for the explanation. This makes sense to me. 
Side question: Should i move the `urlCacher` ( or `getItemPropertyer`) into `BackupHandler`? Seems like a natural fit since the url cache is only relevant during backups.